configuration/postgres/postgres.go

```diff
@@ -0,0 +1,401 @@
+	}
+}
+
+func (p *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	p.logger.Debug(InfoStartInit)
+
+	if p.client != nil {
+		return fmt.Errorf(ErrorAlreadyInitialized)
+	}
+	if m, err := parseMetadata(metadata); err != nil {
+		p.logger.Error(err)
+		return err
+	} else {
+		p.metadata = m
+	}
+	p.ActiveSubscriptions = make(map[string]*subscription)
+	ctx, cancel := context.WithTimeout(context.Background(), p.metadata.maxIdleTime)
+	defer cancel()
+	client, err := Connect(ctx, p.metadata.connectionString, p.metadata.maxIdleTime)
+	if err != nil {
+		return err
+	}
+	p.client = client
+	pingErr := p.client.Ping(ctx)
+	if pingErr != nil {
+		return pingErr
+	}
+	// check if table exists
+	exists := false
+	err = p.client.QueryRow(ctx, QueryTableExists, p.metadata.configTable).Scan(&exists)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequest) (*configuration.GetResponse, error) {
+	if err := validateInput(req.Keys); err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	query, params, err := buildQuery(req, p.metadata.configTable)
+	if err != nil {
+		p.logger.Error(err)
+		return nil, err
+	}
+	rows, err := p.client.Query(ctx, query, params...)
+	if err != nil {
+		// If no rows exist, return an empty response, otherwise return the error.
+		if err == sql.ErrNoRows {
+			return &configuration.GetResponse{}, nil
+		}
+		return nil, err
+	}
+	items := make(map[string]*configuration.Item)
+	for i := 0; rows.Next(); i++ {
+		var item configuration.Item
+		var key string
+		var metadata []byte
+		v := make(map[string]string)
+
+		if err := rows.Scan(&key, &item.Value, &item.Version, &metadata); err != nil {
+			return nil, err
+		}
+		if err := json.Unmarshal(metadata, &v); err != nil {
+			return nil, err
+		}
+		item.Metadata = v
+		if item.Value != "" {
+			items[key] = &item
+		}
+	}
+	return &configuration.GetResponse{
+		Items: items,
+	}, nil
+}
+
+func (p *ConfigurationStore) Subscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler) (string, error) {
+	var triggers []string
+	for k, v := range req.Metadata {
+		if res := strings.EqualFold("trigger", k); res {
+			triggers = append(triggers, v)
+		}
+	}
+
+	if len(triggers) == 0 {
+		return "", fmt.Errorf("cannot subscribe to empty trigger")
+	}
+
+	var subscribeID string
+	for _, trigger := range triggers {
+		key := "listen " + trigger
+		if sub, isActive := p.isSubscriptionActive(req); isActive {
+			if oldStopChan, ok := p.subscribeStopChanMap.Load(sub); ok {
+				close(oldStopChan.(chan struct{}))
+			}
+		}
+		stop := make(chan struct{})
+		subscribeID = uuid.New().String()
+		p.subscribeStopChanMap.Store(subscribeID, stop)
+		p.ActiveSubscriptions[trigger] = &subscription{
+			uuid:    subscribeID,
+			trigger: trigger,
+			keys:    req.Keys,
+		}
+		go p.doSubscribe(ctx, req, handler, key, trigger, subscribeID, stop)
```

Comment:
This method gets a new connection from the pool for each key. Would be better to have a single connection and multiple listeners attached to that? Otherwise, with N keys, you'd need N connections.

Reply:
The key here is not the individual key, but the trigger. I have renamed this variable to make it obvious - https://github.com/dapr/components-contrib/blob/4f64eefae64507ca5cfd5f11044283816b126d4a/configuration/postgres/postgres.go#L167

One channel is created per trigger and when the same trigger is resubscribed to, the existing channel is closed and a new channel is created.

Reply:
Yes, my concern is with the usage of pgx connections. Right now, `doSubscribe` gets a new connection from the pool for each key you're subscribed to. If you subscribe to 10 connections, that requires 10 active connections. Most databases have a limit on # of active connections, so you'd be exhausting them quickly. A single connection should allow `listen`'ing to multiple keys

Reply:
If you look at the client attribute, you will see that it is a pgxpool object. Connection can be reused only when it is closed. So we can reuse connection for get operation. But for subscribe we defer the connection close until unsubscribe. Connection gives a connection busy error when reused, since it is still in use. When we donot mention the number of connections, the default connections in pool is 5. User will not be able to subscribe beyond 5 triggers. Since the need is to listen to notifications from one table only, this should not be a blocker. Will add a new maxconnections to input arguments so that user is not limited by the default. This will also be updated in the docs.

Reply:
Hi @akhilac1 I understand how the pooling works. What I'm saying is that here, you are using one connection for each LISTEN. That seems wasteful, as I believe you should be able to perform multiple LISTEN on a single connection.

Each connection uses quite a bit of resources, both on the client and server (servers too have a connection limit, for example on Azure Postgres it can be as low as 50 on the server).

I think you should have a single, persistent connection that is dedicated to LISTEN'ing, and just add or remove listeners to that connection rather than creating new ones.

Reply:
@ItalyPaleAle - one connection allows subscription to one channel - one channel allows listening to notifications for any number of keys. Added guidance in documentation to use one subscription with multiple keys

_When calling subscribe API, metadata.pgNotifyChannel should be used to specify the name of the channel to listen for notifications from postgres configuration store.

Any number of keys can be added to a subscription request. Each subscription uses a exclusive database connectio. So, it is strongly recommended to subscribe to multiple keys within a single subscription. This helps optimize the number of connections to the database._

https://github.com/dapr/docs/blob/357ffcf6dd53d5736461c7e63050da5917aa6bbf/daprdocs/content/en/reference/components-reference/supported-configuration-stores/postgres-configuration-store.md